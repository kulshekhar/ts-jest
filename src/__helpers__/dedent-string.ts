export const omitLeadingWhitespace = (text: string): string => {
  return text.replace(/^\s+/gm, '')
}

export const dedent = (strings: TemplateStringsArray, ...values: unknown[]) => {
  let joinedString = ''
  for (let i = 0; i < values.length; i++) {
    joinedString += `${strings[i]}${values[i]}`
  }
  joinedString += strings[strings.length - 1]

  return omitLeadingWhitespace(joinedString)
}
