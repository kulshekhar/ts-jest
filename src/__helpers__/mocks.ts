import { testing } from 'bs-logger'

import { rootLogger } from '../util/logger'

export const logTargetMock = () => (rootLogger as testing.LoggerMock).target

export const mockObject = <T, M>(obj: T, newProps: M): T & M & { mockRestore: () => T } => {
  const backup: any = Object.create(null)

  Object.keys(newProps).forEach(key => {
    const desc = (backup[key] = Object.getOwnPropertyDescriptor(obj, key))
    const newDesc: any = { ...desc }
    if (newDesc.get) {
      newDesc.get = () => (newProps as any)[key]
    } else {
      newDesc.value = (newProps as any)[key]
    }
    Object.defineProperty(obj, key, newDesc)
  })
  if ((obj as any).mockRestore) backup.mockRestore = Object.getOwnPropertyDescriptor(obj, 'mockRestore')
  return Object.defineProperty(obj, 'mockRestore', {
    value() {
      Object.keys(backup).forEach(key => {
        Object.defineProperty(obj, key, backup[key])
      })
      return obj
    },
    configurable: true,
  })
}

export const mockWriteStream = () => ({
  written: [] as string[],
  write(text: string) {
    this.written.push(text)
  },
  clear() {
    this.written = []
  },
})
