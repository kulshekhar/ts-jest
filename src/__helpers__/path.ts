import { ensureDirSync } from 'fs-extra'
import { tmpdir } from 'os'
import { join, resolve } from 'path'

export const ROOT = resolve(__dirname, '..', '..')

export function tempDir(ns: string): string {
  const dir = join(tmpdir(), 'ts-jest-unit-tests', ns, parseInt((Math.random() * 1e17) as any, 10).toString(36))
  ensureDirSync(dir)

  return dir
}
