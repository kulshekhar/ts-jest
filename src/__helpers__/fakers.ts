import { Config } from '@jest/types'
import { resolve } from 'path'

import { createCompilerInstance } from '../compiler/instance'
import { ConfigSet } from '../config/config-set'
import { BabelConfig, TsCompiler, TsJestConfig, TsJestGlobalOptions } from '../types'
import { ImportReasons } from '../util/messages'

export function filePath(relPath: string): string {
  return resolve(__dirname, '..', '..', relPath)
}

export const rootDir = filePath('')

export function tsJestConfig(options?: Partial<TsJestConfig>): TsJestConfig {
  return {
    isolatedModules: false,
    compiler: 'typescript',
    transformers: [],
    babelConfig: undefined,
    tsConfig: undefined,
    packageJson: undefined,
    stringifyContentPathRegex: undefined,
    diagnostics: { ignoreCodes: [], pretty: false, throws: true },
    ...options,
  }
}

function getJestConfig<T extends Config.ProjectConfig>(
  options?: Partial<Config.InitialOptions | Config.ProjectConfig>,
  tsJestOptions?: TsJestGlobalOptions,
): T {
  const res = {
    globals: {},
    ...options,
  } as any
  if (tsJestOptions) {
    res.globals['ts-jest'] = tsJestOptions
  }

  return res
}

export function babelConfig<T extends BabelConfig>(options?: BabelConfig): T {
  return {
    ...options,
    presets: [...(options && options.presets)],
    plugins: [...(options && options.plugins)],
  } as any
}

export function importReason(text = '[[BECAUSE]]'): ImportReasons {
  return text as any
}

export const defaultResolve = (path: string): string => `resolved:${path}`

export function createConfigSet({
  jestConfig,
  tsJestConfig,
  parentConfig,
  resolve = defaultResolve,
  ...others
}: {
  jestConfig?: Partial<Config.ProjectConfig>
  tsJestConfig?: TsJestGlobalOptions
  parentConfig?: TsJestGlobalOptions
  resolve?: ((path: string) => string) | null
  [key: string]: any
} = {}): ConfigSet {
  const defaultTestRegex = ['(/__tests__/.*|(\\\\.|/)(test|spec))\\\\.[jt]sx?$']
  const defaultTestMatch = ['**/__tests__/**/*.[jt]s?(x)', '**/?(*.)+(spec|test).[jt]s?(x)']
  jestConfig = {
    ...jestConfig,
    testMatch: jestConfig?.testMatch ? [...jestConfig.testMatch, ...defaultTestMatch] : defaultTestMatch,
    testRegex: jestConfig?.testRegex ? [...defaultTestRegex, ...jestConfig.testRegex] : defaultTestRegex,
  }
  const cs = new ConfigSet(getJestConfig(jestConfig, tsJestConfig), parentConfig)
  if (resolve) {
    cs.resolvePath = resolve
  }
  Object.keys(others).forEach((key) => {
    Object.defineProperty(cs, key, { value: others[key] })
  })

  return cs
}

// not really unit-testing here, but it's hard to mock all those values :-D
export function makeCompiler({
  jestConfig,
  tsJestConfig,
  parentConfig,
}: {
  jestConfig?: Partial<Config.ProjectConfig>
  tsJestConfig?: TsJestGlobalOptions
  parentConfig?: TsJestGlobalOptions
} = {}): TsCompiler {
  tsJestConfig = { ...tsJestConfig }
  tsJestConfig.diagnostics = {
    ...(tsJestConfig.diagnostics as any),
    pretty: false,
  }
  const cs = createConfigSet({ jestConfig, tsJestConfig, parentConfig, resolve: null })

  return createCompilerInstance(cs)
}
