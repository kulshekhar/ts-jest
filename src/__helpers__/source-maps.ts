// WARNING: this file is shared between e2e and unit tests
import bufferFrom = require('buffer-from')
import stableStringify = require('fast-json-stable-stringify')
import { realpathSync } from 'fs'
import { isAbsolute, relative } from 'path'
import { RawSourceMap } from 'source-map'

export function base64ToSourceMaps(base64: string): RawSourceMap {
  return JSON.parse(bufferFrom(base64, 'base64').toString('utf8'))
}

export function sourceMapsToBase64(sourceMaps: RawSourceMap): string {
  return bufferFrom(stableStringify(sourceMaps)).toString('base64')
}

export interface ParsedSourceWithMaps {
  sourceMaps?: RawSourceMap
  comment?: string
  source: string
}
export function parseSource(source: string): ParsedSourceWithMaps {
  const [comment, b64Maps]: [string, string | undefined] =
    // eslint-disable-next-line @typescript-eslint/prefer-regexp-exec
    (source.match(/[\n^]\/\/#\s*sourceMappingURL=data:application\/json;(?:charset=utf-8;)?base64,(\S+)\s*$/) as any) ||
    []
  if (b64Maps) {
    const map = base64ToSourceMaps(b64Maps)
    return {
      source: source.substr(0, -comment.length),
      comment: comment.trim(),
      sourceMaps: map,
    }
  }
  return { source }
}

// source-map module doesn't provide a sync method to extract the source-maps
export function extractSourceMaps(source: string): RawSourceMap | undefined {
  return parseSource(source).sourceMaps
}

export function relativisePaths(map: RawSourceMap, fromPath: string, newPrefix = ''): RawSourceMap {
  const res = { ...map }
  const from = realpathSync(fromPath)
  const remap = (path: string): string =>
    (isAbsolute(path) ? `${newPrefix}${relative(from, realpathSync(path))}` : path).replace(/\\/g, '/')
  if (res.sourceRoot) res.sourceRoot = remap(res.sourceRoot)
  if (res.sources) res.sources = res.sources.map(remap)
  if (res.file) res.file = remap(res.file)
  return res
}

export function rewriteSourceMaps(source: string, sourceMapsTransformer: (maps: RawSourceMap) => RawSourceMap): string {
  return source.replace(
    /([\n^]\/\/#\s*sourceMappingURL=data:application\/json;(?:charset=utf-8;)?base64,)(\S+)(\s*)$/,
    (_, before, base64, after) => {
      let map = base64ToSourceMaps(base64)
      map = sourceMapsTransformer(map)
      return `${before}${sourceMapsToBase64(map)}${after}`
    },
  )
}
