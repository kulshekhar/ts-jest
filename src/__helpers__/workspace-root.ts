import fs from 'node:fs'
import path from 'node:path'

function workspaceRootInner(dir: string, candidateRoot: string) {
  if (path.dirname(dir) === dir) return candidateRoot

  const matchablePaths = [path.join(dir, '.ts-jest-digest'), path.join(dir, '.github'), path.join(dir, 'renovate.json')]
  if (matchablePaths.some((x) => fs.existsSync(x))) {
    return dir
  }

  return workspaceRootInner(path.dirname(dir), candidateRoot)
}

export const workspaceRoot = workspaceRootInner(process.cwd(), process.cwd())
