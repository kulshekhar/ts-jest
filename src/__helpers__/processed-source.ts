import { isAbsolute, relative } from 'path'
import { RawSourceMap } from 'source-map'

import { ROOT } from './path'
import { ParsedSourceWithMaps, parseSource, relativisePaths, rewriteSourceMaps } from './source-maps'

export default class ProcessedSource {
  readonly filename: string

  constructor(readonly output: string, filename: string, readonly cwd: string = ROOT) {
    if (isAbsolute(filename)) filename = relative(this.cwd, filename)
    this.filename = filename
  }
  get outputCode(): string {
    return this.output
  }
  get parsedSource(): ParsedSourceWithMaps {
    return parseSource(this.output)
  }
  get outputCodeWithoutMaps(): string {
    return this.parsedSource.source
  }
  get outputSourceMaps(): RawSourceMap | undefined {
    return this.parsedSource.sourceMaps
  }
  get normalizedOutputCode(): string {
    return rewriteSourceMaps(this.output, this.sourceMapsNormalizer)
  }
  get normalizedOutputSourceMaps(): RawSourceMap | undefined {
    const maps = this.outputSourceMaps
    if (maps) return this.sourceMapsNormalizer(maps)

    return undefined
  }
  get sourceMapsNormalizer() {
    return (maps: RawSourceMap): RawSourceMap => relativisePaths(maps, this.cwd, '<cwd>/')
  }
}
