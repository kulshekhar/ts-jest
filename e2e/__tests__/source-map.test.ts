import { join } from 'path'

import { allValidPackageSets } from '../__helpers__/templates'
import { configureTestCase } from '../__helpers__/test-case'

describe('Source map', () => {
  const testCase = configureTestCase('source-maps', {
    writeIo: true,
    // TS5023 - unrecognized compiler option
    // TS7027 - unreachable code
    tsJestConfig: { diagnostics: { ignoreCodes: [5023, 7027] } },
  })

  testCase.runWithTemplates(allValidPackageSets, 1, (runTest, { templateName }) => {
    describe(`using template "${templateName}"`, () => {
      const result = runTest()
      it(`should report correct line numbers`, () => {
        expect(result.status).toBe(1)
        expect(result).toMatchSnapshot()
      })

      // when there are some issues with debugging, it's usually because source maps are not inlined
      // and the debugger cannot find the line where to go
      it(`should have the source maps comment`, () => {
        const io = result.ioFor('main.ts')
        const filePath = join(result.cwd, 'main.ts')
        expect(io).toMatchSnapshot()
        // we could keep only the snapshot testing, but this ensure fixed stuff are the same over versions
        expect(io.outputSourceMaps).toMatchObject({
          names: [],
          sources: [filePath],
          sourcesContent: [io.inputCode],
          version: 3,
        })
      })
    })
  })
})
