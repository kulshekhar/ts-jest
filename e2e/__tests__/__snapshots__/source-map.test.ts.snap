// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`using template "default" should have the source maps comment 1`] = `
  ===[ FILE: main.ts ]============================================================
  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  function consoleLog() {
      console.log('WITHIN SOURCE');
  }
  exports.consoleLog = consoleLog;
  function throwError() {
      throw new Error('WITHIN SOURCE');
  }
  exports.throwError = throwError;
  //# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJmaWxlIjoiPGN3ZD4vbWFpbi50cyIsIm1hcHBpbmdzIjoiOztBQUFBLFNBQWdCLFVBQVU7SUFDeEIsT0FBTyxDQUFDLEdBQUcsQ0FBQyxlQUFlLENBQUMsQ0FBQTtBQUM5QixDQUFDO0FBRkQsZ0NBRUM7QUFFRCxTQUFnQixVQUFVO0lBQ3hCLE1BQU0sSUFBSSxLQUFLLENBQUMsZUFBZSxDQUFDLENBQUE7QUFDbEMsQ0FBQztBQUZELGdDQUVDIiwibmFtZXMiOltdLCJzb3VyY2VzIjpbIjxjd2Q+L21haW4udHMiXSwic291cmNlc0NvbnRlbnQiOlsiZXhwb3J0IGZ1bmN0aW9uIGNvbnNvbGVMb2coKSB7XG4gIGNvbnNvbGUubG9nKCdXSVRISU4gU09VUkNFJylcbn1cblxuZXhwb3J0IGZ1bmN0aW9uIHRocm93RXJyb3IoKSB7XG4gIHRocm93IG5ldyBFcnJvcignV0lUSElOIFNPVVJDRScpXG59XG4iXSwidmVyc2lvbiI6M30=
  ===[ INLINE SOURCE MAPS ]=======================================================
  file: <cwd>/main.ts
  mappings: >-
    ;;AAAA,SAAgB,UAAU;IACxB,OAAO,CAAC,GAAG,CAAC,eAAe,CAAC,CAAA;AAC9B,CAAC;AAFD,gCAEC;AAED,SAAgB,UAAU;IACxB,MAAM,IAAI,KAAK,CAAC,eAAe,CAAC,CAAA;AAClC,CAAC;AAFD,gCAEC
  names: []
  sources:
    - <cwd>/main.ts
  sourcesContent:
    - |
      export function consoleLog() {
        console.log('WITHIN SOURCE')
      }
  
      export function throwError() {
        throw new Error('WITHIN SOURCE')
      }
  version: 3
  ================================================================================
`;

exports[`using template "default" should report correct line numbers 1`] = `
  × jest --no-cache
  ↳ exit code: 1
  ===[ STDOUT ]===================================================================
  console.log main.ts:2
      WITHIN SOURCE
  
    console.log main.spec.ts:9
      WITHIN TEST
  ===[ STDERR ]===================================================================
  FAIL ./main.spec.ts
    console.log()
      √ from sources
      √ from tests
    throw new Error()
      × throws from sources
      × throws from tests
  
    ● throw new Error() › throws from sources
  
      WITHIN SOURCE
  
        4 | 
        5 | export function throwError() {
      > 6 |   throw new Error('WITHIN SOURCE')
          |         ^
        7 | }
        8 | 
  
        at Object.throwError (main.ts:6:9)
        at Object.<anonymous> (main.spec.ts:16:5)
  
    ● throw new Error() › throws from tests
  
      WITHIN TEST
  
        18 |   });
        19 |   test('throws from tests', () => {
      > 20 |     throw new Error('WITHIN TEST');
           |           ^
        21 |     expect(true).toBe(true);
        22 |   });
        23 | });
  
        at Object.<anonymous> (main.spec.ts:20:11)
  
  Test Suites: 1 failed, 1 total
  Tests:       2 failed, 2 passed, 4 total
  Snapshots:   0 total
  Time:        XXs
  Ran all test suites.
  ================================================================================
`;

exports[`using template "with-babel-7" should have the source maps comment 1`] = `
  ===[ FILE: main.ts ]============================================================
  "use strict";
  
  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  
  function consoleLog() {
    console.log('WITHIN SOURCE');
  }
  
  exports.consoleLog = consoleLog;
  
  function throwError() {
    throw new Error('WITHIN SOURCE');
  }
  
  exports.throwError = throwError;
  //# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJtYXBwaW5ncyI6Ijs7Ozs7O0FBQUEsU0FBZ0IsVUFBaEIsR0FBMEI7QUFDeEIsRUFBQSxPQUFPLENBQUMsR0FBUixDQUFZLGVBQVo7QUFDRDs7QUFGRCxPQUFBLENBQUEsVUFBQSxHQUFBLFVBQUE7O0FBSUEsU0FBZ0IsVUFBaEIsR0FBMEI7QUFDeEIsUUFBTSxJQUFJLEtBQUosQ0FBVSxlQUFWLENBQU47QUFDRDs7QUFGRCxPQUFBLENBQUEsVUFBQSxHQUFBLFVBQUEiLCJuYW1lcyI6W10sInNvdXJjZXMiOlsiPGN3ZD4vbWFpbi50cyJdLCJzb3VyY2VzQ29udGVudCI6WyJleHBvcnQgZnVuY3Rpb24gY29uc29sZUxvZygpIHtcbiAgY29uc29sZS5sb2coJ1dJVEhJTiBTT1VSQ0UnKVxufVxuXG5leHBvcnQgZnVuY3Rpb24gdGhyb3dFcnJvcigpIHtcbiAgdGhyb3cgbmV3IEVycm9yKCdXSVRISU4gU09VUkNFJylcbn1cbiJdLCJ2ZXJzaW9uIjozfQ==
  ===[ INLINE SOURCE MAPS ]=======================================================
  mappings: >-
    ;;;;;;AAAA,SAAgB,UAAhB,GAA0B;AACxB,EAAA,OAAO,CAAC,GAAR,CAAY,eAAZ;AACD;;AAFD,OAAA,CAAA,UAAA,GAAA,UAAA;;AAIA,SAAgB,UAAhB,GAA0B;AACxB,QAAM,IAAI,KAAJ,CAAU,eAAV,CAAN;AACD;;AAFD,OAAA,CAAA,UAAA,GAAA,UAAA
  names: []
  sources:
    - <cwd>/main.ts
  sourcesContent:
    - |
      export function consoleLog() {
        console.log('WITHIN SOURCE')
      }
  
      export function throwError() {
        throw new Error('WITHIN SOURCE')
      }
  version: 3
  ================================================================================
`;

exports[`using template "with-babel-7" should report correct line numbers 1`] = `
  × jest --no-cache
  ↳ exit code: 1
  ===[ STDOUT ]===================================================================
  console.log main.ts:2
      WITHIN SOURCE
  
    console.log main.spec.ts:9
      WITHIN TEST
  ===[ STDERR ]===================================================================
  FAIL ./main.spec.ts
    console.log()
      √ from sources
      √ from tests
    throw new Error()
      × throws from sources
      × throws from tests
  
    ● throw new Error() › throws from sources
  
      WITHIN SOURCE
  
        4 | 
        5 | export function throwError() {
      > 6 |   throw new Error('WITHIN SOURCE')
          |         ^
        7 | }
        8 | 
  
        at Object.throwError (main.ts:6:9)
        at Object.<anonymous> (main.spec.ts:16:5)
  
    ● throw new Error() › throws from tests
  
      WITHIN TEST
  
        18 |   });
        19 |   test('throws from tests', () => {
      > 20 |     throw new Error('WITHIN TEST');
           |           ^
        21 |     expect(true).toBe(true);
        22 |   });
        23 | });
  
        at Object.<anonymous> (main.spec.ts:20:11)
  
  Test Suites: 1 failed, 1 total
  Tests:       2 failed, 2 passed, 4 total
  Snapshots:   0 total
  Time:        XXs
  Ran all test suites.
  ================================================================================
`;

exports[`using template "with-typescript-2-7" should have the source maps comment 1`] = `
  ===[ FILE: main.ts ]============================================================
  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  function consoleLog() {
      console.log('WITHIN SOURCE');
  }
  exports.consoleLog = consoleLog;
  function throwError() {
      throw new Error('WITHIN SOURCE');
  }
  exports.throwError = throwError;
  //# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJmaWxlIjoiPGN3ZD4vbWFpbi50cyIsIm1hcHBpbmdzIjoiOztBQUFBO0lBQ0UsT0FBTyxDQUFDLEdBQUcsQ0FBQyxlQUFlLENBQUMsQ0FBQTtBQUM5QixDQUFDO0FBRkQsZ0NBRUM7QUFFRDtJQUNFLE1BQU0sSUFBSSxLQUFLLENBQUMsZUFBZSxDQUFDLENBQUE7QUFDbEMsQ0FBQztBQUZELGdDQUVDIiwibmFtZXMiOltdLCJzb3VyY2VzIjpbIjxjd2Q+L21haW4udHMiXSwic291cmNlc0NvbnRlbnQiOlsiZXhwb3J0IGZ1bmN0aW9uIGNvbnNvbGVMb2coKSB7XG4gIGNvbnNvbGUubG9nKCdXSVRISU4gU09VUkNFJylcbn1cblxuZXhwb3J0IGZ1bmN0aW9uIHRocm93RXJyb3IoKSB7XG4gIHRocm93IG5ldyBFcnJvcignV0lUSElOIFNPVVJDRScpXG59XG4iXSwidmVyc2lvbiI6M30=
  ===[ INLINE SOURCE MAPS ]=======================================================
  file: <cwd>/main.ts
  mappings: >-
    ;;AAAA;IACE,OAAO,CAAC,GAAG,CAAC,eAAe,CAAC,CAAA;AAC9B,CAAC;AAFD,gCAEC;AAED;IACE,MAAM,IAAI,KAAK,CAAC,eAAe,CAAC,CAAA;AAClC,CAAC;AAFD,gCAEC
  names: []
  sources:
    - <cwd>/main.ts
  sourcesContent:
    - |
      export function consoleLog() {
        console.log('WITHIN SOURCE')
      }
  
      export function throwError() {
        throw new Error('WITHIN SOURCE')
      }
  version: 3
  ================================================================================
`;

exports[`using template "with-typescript-2-7" should report correct line numbers 1`] = `
  × jest --no-cache
  ↳ exit code: 1
  ===[ STDOUT ]===================================================================
  console.log main.ts:2
      WITHIN SOURCE
  
    console.log main.spec.ts:9
      WITHIN TEST
  ===[ STDERR ]===================================================================
  FAIL ./main.spec.ts
    console.log()
      √ from sources
      √ from tests
    throw new Error()
      × throws from sources
      × throws from tests
  
    ● throw new Error() › throws from sources
  
      WITHIN SOURCE
  
        4 | 
        5 | export function throwError() {
      > 6 |   throw new Error('WITHIN SOURCE')
          |         ^
        7 | }
        8 | 
  
        at Object.throwError (main.ts:6:9)
        at Object.<anonymous> (main.spec.ts:16:5)
  
    ● throw new Error() › throws from tests
  
      WITHIN TEST
  
        18 |   });
        19 |   test('throws from tests', () => {
      > 20 |     throw new Error('WITHIN TEST');
           |           ^
        21 |     expect(true).toBe(true);
        22 |   });
        23 | });
  
        at Object.<anonymous> (main.spec.ts:20:11)
  
  Test Suites: 1 failed, 1 total
  Tests:       2 failed, 2 passed, 4 total
  Snapshots:   0 total
  Time:        XXs
  Ran all test suites.
  ================================================================================
`;
